src/coreComponents/physicsSolvers/fluidFlow/OBLSuperEngineKernels.hpp

```diff
@@ -0,0 +1,1507 @@
+
+    localIndex const erJ  = m_seri( iconn, 1 );
+    localIndex const esrJ = m_sesri( iconn, 1 );
+    localIndex const eiJ  = m_sei( iconn, 1 );
+
+    arraySlice1d< real64 const, compflow::USD_OBL_VAL - 1 > const & OBLValsI = m_OBLOperatorValues[erI][esrI][eiI];
+    arraySlice2d< real64 const, compflow::USD_OBL_DER - 1 > const & OBLDersI = m_OBLOperatorDerivatives[erI][esrI][eiI];
+
+    arraySlice1d< real64 const, compflow::USD_OBL_VAL - 1 > const & OBLValsJ = m_OBLOperatorValues[erJ][esrJ][eiJ];
+    arraySlice2d< real64 const, compflow::USD_OBL_DER - 1 > const & OBLDersJ = m_OBLOperatorDerivatives[erJ][esrJ][eiJ];
+
+    // single gravity coefficient and transimissibility values correspond to the current connection
+    real64 const gravCoef = m_gravCoef[erI][esrI][eiI] - m_gravCoef[erJ][esrJ][eiJ];
+    real64 const trans = stack.transmissibility[0][0] * transUnitMult;
+    real64 const transD = stack.diffusiveTransmissibility[0][0];
+
+
+    real64 transMult = 1;
+    real64 transMultDerI[numDofs];
+    real64 transMultDerJ[numDofs];
+    if( m_transMultExp > 0 )
+    {
+      // Calculate transmissibility multiplier:
+
+      // Take average interface porosity:
+      real64 const poroAverage = (OBLValsI[PORO_OP] + OBLValsJ[PORO_OP]) * 0.5;
+      real64 const transMultD = m_transMultExp * pow( poroAverage, m_transMultExp - 1 ) * 0.5;
+
+      for( integer v = 0; v < numDofs; v++ )
+      {
+        transMultDerI[v] = transMultD * OBLDersI[PORO_OP][v];
+        transMultDerJ[v] = transMultD * OBLDersJ[PORO_OP][v];
+      }
+      transMult = pow( poroAverage, m_transMultExp );
+    }
+    else
+    {
+      for( integer v = 0; v < numDofs; v++ )
+      {
+        transMultDerI[v] = 0;
+        transMultDerJ[v] = 0;
+      }
+    }
+
+    real64 const pDiff = m_pres[erJ][esrJ][eiJ] + m_dPres[erJ][esrJ][eiJ] - m_pres[erI][esrI][eiI] - m_dPres[erI][esrI][eiI];
+
+
+    // [2] fill offdiagonal part + contribute to diagonal, only fluid part is considered in energy equation
+    for( integer p = 0; p < numPhases; p++ )
+    {     // loop over number of phases for convective operator
+
+      // calculate gravity term for phase p
+      real64 const averageDensity = (OBLValsI[GRAV_OP + p] + OBLValsJ[GRAV_OP + p]) / 2;
+
+      // p = 1 means oil phase, it's reference phase. pw=po-pcow, pg=po-(-pcog).
+      real64 const phasePDiff = (pDiff + averageDensity * gravCoef - OBLValsJ[PC_OP + p] + OBLValsI[PC_OP + p]) * pascalToBarMult;
+
+      // calculate partial derivatives for gravity and capillary terms
+      real64 gravPcDerI[numDofs];
+      real64 gravPcDerJ[numDofs];
+      for( integer v = 0; v < numDofs; v++ )
+      {
+        gravPcDerI[v] = -(OBLDersI[GRAV_OP + p][v]) * gravCoef / 2 - OBLDersI[PC_OP + p][v];
+        gravPcDerJ[v] = -(OBLDersJ[GRAV_OP + p][v]) * gravCoef / 2 + OBLDersJ[PC_OP + p][v];
+      }
+
+      real64 phaseGammaPDiff = transMult * trans * m_dt * phasePDiff;
+
+      if( phasePDiff < 0 )
```

Comment:
To simplify the upwinding, do you think it would be possible to create the following variables:
```OBLValsUp = m_OBLOperatorValues[erUp][esrUp][eiUp];```
and
```OBLDersUp = m_OBLOperatorDerivatives[erUp][esrUp][eiUp];```
where the indices `erUp`, `esrUp`, and `eiUp` would be deduced from the sign of the potential difference? 

That would allow you to have only one version of the code that is currently in the `if ... else` just below.

The same simplification could probably be done for the diffusion flux and energy flux at the end of this function, if I understand correctly.

Reply:
Great idea, I like how it looks now. Hopefully I did not introduce (new) bugs here, existing unit tests do not provide reliable confirmation...